src/upgrades/v0_9_0/ca_objects_migration.rs

```diff
@@ -145,109 +146,102 @@ impl UpgradeStore for CasStoreMigration {
         unreachable!("checked directly on keystore")
     }
 
-    fn migrate(&self) -> Result<(), UpgradeError> {
-        info!("Krill will now reformat existing command and event data, and remove unnecessary data");
+    fn prepare_new_data(&self, mode: UpgradeMode) -> Result<(), UpgradeError> {
+        info!("Upgrade CA command and event data to Krill version {}", KRILL_VERSION);
 
         let dflt_actor = "krill".to_string();
 
         // For each CA:
-        //   - build a new
-        for scope in self.store.scopes()? {
-            // Archive all keys in the scope, then we can write new keys as needed without
-            // overwriting anything when we renumber.
-            for key in self.store.keys(Some(scope.clone()), "")? {
-                self.archive_to_migration_scope(&key)?;
-            }
-
-            let migration_scope = format!("{}/{}", scope, MIGRATION_SCOPE);
-
-            let migration_info_key = KeyStoreKey::scoped(migration_scope.clone(), "info.json".to_string());
-            let mut info: StoredValueInfo = match self.store.get(&migration_info_key) {
-                Ok(Some(info)) => info,
-                _ => StoredValueInfo::default(),
-            };
+        for scope in self.current_kv_store.scopes()? {
+            // Getting the Handle should never fail, but if it does then we should bail out asap.
+            let handle = Handle::from_str(&scope)
+                .map_err(|_| UpgradeError::Custom(format!("Found invalid CA handle '{}'", scope)))?;
+
+            // Get the info from the current store to see where we are
+            let mut data_upgrade_info = self.data_upgrade_info(&scope)?;
+
+            // Get the list of commands to prepare, starting with the last_command we got to (may be 0)
+            let old_cmd_keys = self.command_keys(&scope, data_upgrade_info.last_command)?;
+
+            // Migrate the initialisation event, if not done in a previous run. This
+            // is a special event that has no command, so we need to do this separately.
+            if data_upgrade_info.last_event == 0 {
+                // Make a new init event.
+                let init_key = Self::event_key(&scope, 0);
+
+                let old_init: OldCaIni = self.get(&init_key)?;
+                let (id, _, old_ini_det) = old_init.unpack();
+                let (rfc_8183_id, repo_opt, ta_opt) = old_ini_det.unpack();
+                let ini = IniDet::new(&id, rfc_8183_id.into());
+                self.new_kv_store.store(&init_key, &ini)?;
+
+                // If the CA was initialized as a trust anchor, then we refuse to upgrade.
+                // This can only happen if this is a very old test system. People will need
+                // to set up a new test system instead.
+                if ta_opt.is_some() {
```

Comment:
If we are going to refuse to upgrade should we have already added the init even then?

Reply:
along the way, I needed it for when prepping the data for the upgrade.